k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -24,11 +25,28 @@ class EndpointsNamer(
    * and attempts to bind an Addr by resolving named endpoint from the
    * kubernetes master.
    */
-  def lookup(path: Path): Activity[NameTree[Name]] = path.take(PrefixLen) match {
-    case id@Path.Utf8(nsName, portName, serviceName) =>
-      val residual = path.drop(PrefixLen)
+  def lookup(path: Path): Activity[NameTree[Name]] = (path.take(variablePrefixLength), labelName) match {
+    case (id@Path.Utf8(nsName, _, _), None) =>
+      val residual = path.drop(variablePrefixLength)
       log.debug("k8s lookup: %s %s", id.show, path.show)
-      endpointNs.get(nsName).services.flatMap { services =>
+      lookupServices(endpointNs.get(nsName, None), id, residual)
+
+    case (id@Path.Utf8(nsName, _, _, labelValue), Some(label)) =>
+      val residual = path.drop(variablePrefixLength)
+      log.debug("k8s lookup: %s %s %s", id.show, label, path.show)
+      lookupServices(endpointNs.get(nsName, Some(s"${label}=${labelValue}")), id.take(PrefixLen), residual)
+
+    case (id@Path.Utf8(nsName, portName, serviceName), Some(label)) =>
+      log.debug("k8s lookup: ns %s service %s label value segment missing for label %s", nsName, serviceName, portName, label)
+      Activity.value(NameTree.Neg)
+
+    case _ =>
+      Activity.value(NameTree.Neg)
+  }
+
+  private[this] def lookupServices(cache: NsCache, id: Path, residual: Path): Activity[NameTree[Name]] = id match {
```

Comment:
It just so happens that every place calling this method has done `id.take(3)` before passing the id in, but I can move the take to be part of this method instead